cpp/src/arrow/array/validate.cc

```diff
@@ -509,22 +510,23 @@ struct ValidateArrayImpl {
     }
 
     if (full_validation) {
-      if (data.null_count != kUnknownNullCount) {
-        int64_t actual_null_count;
-        if (HasValidityBitmap(data.type->id()) && data.buffers[0]) {
-          // Do not call GetNullCount() as it would also set the `null_count` member
-          actual_null_count = data.length - CountSetBits(data.buffers[0]->data(),
-                                                         data.offset, data.length);
-        } else if (data.type->storage_id() == Type::NA) {
-          actual_null_count = data.length;
-        } else {
-          actual_null_count = 0;
-        }
-        if (actual_null_count != data.null_count) {
-          return Status::Invalid("null_count value (", data.null_count,
-                                 ") doesn't match actual number of nulls in array (",
-                                 actual_null_count, ")");
-        }
+      int64_t actual_null_count;
+      if (HasValidityBitmap(data.type->id()) && data.buffers[0]) {
+        // Do not call GetNullCount() as it would also set the `null_count` member
+        actual_null_count =
+            data.length - CountSetBits(data.buffers[0]->data(), data.offset, data.length);
+      } else if (data.type->storage_id() == Type::NA) {
+        actual_null_count = data.length;
+      } else {
+        actual_null_count = 0;
+      }
+      if (data.null_count != kUnknownNullCount && actual_null_count != data.null_count) {
+        return Status::Invalid("null_count value (", data.null_count,
+                               ") doesn't match actual number of nulls in array (",
+                               actual_null_count, ")");
+      }
+      if (!nullable && actual_null_count > 0) {
+        return Status::Invalid("Null found but field is not nullable");
       }
     }
     return Status::OK();
```

Comment:
Also, line 585 below (in ValidateListLike) has a RecurseInto call that should be updated

Reply:
Ah, sorry, somehow I missed that. Thanks